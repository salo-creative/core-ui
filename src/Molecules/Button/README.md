# Button

Renders a styled button or link element, with disabled & loading states and inline iconography

---

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import H1 from '@salo/core-ui/Molecules/Button';
```

Implement as follows

```javascript
<Button>Button text</Button>
```

**For full prop types and usage see storybook info/knobs**