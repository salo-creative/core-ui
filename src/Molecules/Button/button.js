import React from 'react';
import PropTypes from 'prop-types';
import Ink from 'react-ink';
import { Link } from 'react-router-dom';
import Icon from '@salo/icons';

// COMPONENTS
import { ButtonWrapper } from './button.styles';
import Loader from '../Loader';

const Button = (props) => {
  const {
    children,
    circle,
    disabled,
    fullWidth,
    height,
    iconAfter,
    iconBefore,
    loading,
    padding,
    radius,
    type,
    ...rest
  } = props;

  const disable = loading || disabled;

  const evaluateElement = () => {
    switch (type) {
      case 'Link':
        return Link;
      case 'a':
        return 'a';
      case 'button':
      case 'submit':
      default:
        return 'button';
    }
  };

  const renderIconBefore = () => {
    if (!iconBefore) return null;
    return <Icon margin={ children ? '0 1rem 0 0' : '0' } icon={ iconBefore } />;
  };

  const renderIconAfter = () => {
    if (!iconAfter) return null;
    return <Icon margin={ children ? '0 0 0 1rem' : '0' } icon={ iconAfter } />;
  };
  
  return (
    <ButtonWrapper
      { ...rest }
      circle={ circle.toString() }
      disabled={ disable }
      fullwidth={ fullWidth.toString() }
      loading={ loading.toString() }
      radius={ radius.toString() }
      padding={ padding }
      as={ evaluateElement() }
      height={ height }
    >
      <Ink />
      { loading && (
        <Loader
          display={ true }
          loaderProps={ { size: 40, position: 'absolute' } }
          appearance='light'
        />
      ) }
      { renderIconBefore() }
      { children }
      { renderIconAfter() }
    </ButtonWrapper>
  );
};

Button.defaultProps = {
  align: 'center',
  appearance: 'primary',
  children: '',
  circle: false,
  disabled: false,
  fullWidth: false,
  height: '4rem',
  iconAfter: '',
  iconBefore: '',
  loading: false,
  padding: '0 1.5rem',
  radius: true,
  shadow: 'default',
  type: 'button'
};

Button.propTypes = {
  align: PropTypes.oneOf(['center', 'start', 'flex-end', 'space-between']),
  appearance: PropTypes.string,
  children: PropTypes.any,
  circle: PropTypes.bool,
  disabled: PropTypes.bool,
  fullWidth: PropTypes.bool,
  height: PropTypes.string,
  iconAfter: PropTypes.string,
  iconBefore: PropTypes.string,
  loading: PropTypes.bool,
  padding: PropTypes.string,
  radius: PropTypes.bool,
  shadow: PropTypes.oneOf(['small', 'default', 'large', 'none']),
  type: PropTypes.oneOf(['button', 'Link', 'submit', 'a'])
};

export default Button;