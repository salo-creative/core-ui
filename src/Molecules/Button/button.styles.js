import styled from 'styled-components';
import { get } from 'lodash';

// HELPERS
import { buttonThemes, boxShadow } from '../../helpers/colours';

const getColour = ({ props, property }) => {
  return get(buttonThemes(props.theme), `${ props.appearance }.${ property }`);
};

export const ButtonWrapper = styled.button`
  display: inline-flex;
  background-color: ${ props => getColour({ property: 'background', props }) };
  border: 1px solid ${ props => getColour({ property: 'border', props }) };
  color: ${ props => (props.loading === 'true' ? getColour({ property: 'background', props }) : getColour({ property: 'color', props })) };
  transition: all 0.3s linear;
  position: relative;
  justify-content: ${ ({ align }) => align };
  align-items: center;
  cursor: pointer;
  text-decoration: none;
  text-transform: uppercase;
  letter-spacing: 1px;
  border-radius: ${ ({ radius }) => (radius === 'true' ? '0.4rem' : '0') };
  font-size: 1.4rem;
  height: ${ ({ size }) => (size === 'L' ? '4.5rem' : '4rem') };
  padding: 0 1.5rem;
  width: ${ ({ fullwidth }) => (fullwidth === 'true' ? '100%' : 'auto') };
  box-shadow: ${ props => (getColour({ property: 'shadow', props }) ? boxShadow() : 'none') };
  path {
    transition: fill 0.3s linear;
    fill: ${ props => (props.loading === 'true' ? getColour({ property: 'background', props }) : getColour({ property: 'color', props })) } !important;
  }
  &:disabled {
    background-color: ${ props => getColour({ property: 'background', props }) } !important;
    border: 1px solid ${ props => getColour({ property: 'border', props }) };
    color: ${ props => (props.loading === 'true' ? getColour({ property: 'background', props }) : getColour({ property: 'color', props })) };
    opacity: 0.6;
    pointer-events: auto;
    cursor: not-allowed;
    path {
      fill: ${ props => (props.loading === 'true' ? getColour({ property: 'background', props }) : getColour({ property: 'color', props })) } !important;
    }
  }
  &:focus {
    outline: none
  }
  &:hover:not([disabled]) {
    border: 1px solid  ${ props => getColour({ property: 'borderHov', props }) };
    background: ${ props => getColour({ property: 'backgroundHov', props }) };
  }
  &.active,
  &:hover:not([disabled]) {
    color: ${ props => (props.loading === 'true' ? getColour({ property: 'backgroundHov', props }) : getColour({ property: 'colorHov', props })) };
    path {
      fill: ${ props => (props.loading === 'true' ? getColour({ property: 'backgroundHov', props }) : getColour({ property: 'colorHov', props })) } !important;
    }
  }
`;