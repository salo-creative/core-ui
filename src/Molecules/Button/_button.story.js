import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, select, text } from '@storybook/addon-knobs';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Button, Row, Column } from '../../index';

// README //
import README from './README.md';

// Start of story logic
const stories = storiesOf('Molecules | Button', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['button'] });


const columnProps = { tablet: 6, medium: 4, large: 3, style: { border: '1px solid #fff', background: '#f2f3f4' } };

stories.add(
  'Basic',
  (() => {
    // KNOBS
    const appearance = select('appearance', {
      'Primary': 'primary',
      'Secondary': 'secondary',
      'Success': 'success',
      'Error': 'error',
      'White': 'white',
      'Clear': 'clear',
      'Clear Inverse': 'clear_inverse',
      'Text': 'text',
      'Text Inverse': 'text_inverse'
    }, 'primary');
    const type = select('Type', {
      'Button': 'button',
      'Submit': 'submit',
      'Link': 'Link',
      'A': 'a'
    }, 'button');
    const align = select('Align', {
      'Center': 'center',
      'Start': 'start',
      'Flex end': 'flex-end',
      'Space between': 'space-between'
    }, 'center');
    const iconBefore = text('Icon before', 'dashboard');
    const iconAfter = text('Icon after', '');
    const fullWidth = boolean('Display full width', false);
    const loading = boolean('Show loading state', false);
    const disabled = boolean('Show disabled state', false);
    const radius = boolean('Show radius', true);
    const size = select('size', ['L', 'M'], 'M');
    return (
      <React.Fragment>
        <Button
          appearance={ appearance }
          fullWidth={ fullWidth }
          loading={ loading }
          disabled={ disabled }
          iconBefore={ iconBefore }
          iconAfter={ iconAfter }
          radius={ radius }
          align={ align }
          size={ size }
          type={ type }
          to='/'
        >
          Button text
        </Button>
      </React.Fragment>
    );
  }),
  { info: { propTablesExclude: [React.Fragment] }, notes: README }
);

stories.add(
  'All styles',
  (() => {
    // KNOBS
    return (
      <React.Fragment>
        <Row margin='0 -10px 0'>
          <Column { ...columnProps }>
            <Button appearance='primary' fullWidth>Primary button</Button>
          </Column>
          <Column { ...columnProps }>
            <Button appearance='secondary' fullWidth>Secondary button</Button>
          </Column>
          <Column { ...columnProps }>
            <Button appearance='success' fullWidth>Success button</Button>
          </Column>
          <Column { ...columnProps }>
            <Button appearance='error' fullWidth>Error button</Button>
          </Column>
          <Column { ...columnProps }>
            <Button appearance='white' fullWidth>White button</Button>
          </Column>
          <Column { ...columnProps }>
            <Button appearance='clear' fullWidth>Clear button</Button>
          </Column>
          <Column { ...columnProps }>
            <Button appearance='clear_inverse' fullWidth>Clear Inverse button</Button>
          </Column>
          <Column { ...columnProps }>
            <Button appearance='text' fullWidth>Text button</Button>
          </Column>
          <Column { ...columnProps }>
            <Button appearance='text_inverse' fullWidth>Text Inverse button</Button>
          </Column>
        </Row>
      </React.Fragment>
    );
  }),
  { info: { propTablesExclude: [React.Fragment, Column, Row] }, notes: README }
);