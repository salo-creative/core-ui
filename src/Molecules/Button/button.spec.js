import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Button from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Button />
  );
  expect(container.firstChild).toMatchSnapshot();
});