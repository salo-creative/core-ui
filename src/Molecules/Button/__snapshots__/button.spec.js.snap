// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  background-color: #00ACED;
  border: 1px solid #00ACED;
  color: #fff;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  position: relative;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  height: 4rem;
  padding: 0 1.5rem;
  width: auto;
  box-shadow: 2px 2px 2px 0px rgba(0,0,0,0.2);
}

.c0 path {
  -webkit-transition: fill 0.3s linear;
  transition: fill 0.3s linear;
  fill: #fff !important;
}

.c0.disabled,
.c0:disabled {
  background-color: #00ACED !important;
  border: 1px solid #00ACED;
  color: #fff;
  opacity: 0.6;
  pointer-events: auto;
  cursor: not-allowed;
}

.c0.disabled path,
.c0:disabled path {
  fill: #fff !important;
}

.c0:focus {
  outline: none;
}

.c0:hover:not([disabled]) {
  border: 1px solid #0074a1;
  background: #0074a1;
}

.c0.active,
.c0:hover:not([disabled]) {
  color: #fff;
}

.c0.active path,
.c0:hover:not([disabled]) path {
  fill: #fff !important;
}

<button
  class="c0  "
  height="4rem"
  loading="false"
  radius="true"
  type="button"
>
  <canvas
    height="0"
    style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
    width="0"
  />
  
</button>
`;
